java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -83,48 +120,65 @@ class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
-        referenceId: String
+        message: ProxyMessage<Data, Op, T>
     ): Boolean {
-        val (id, store) = store(referenceId)
-        val deMuxedMessage: ProxyMessage<Data, Op, T> = when (message) {
-            is SyncRequest -> SyncRequest(id)
-            is ModelUpdate -> ModelUpdate(message.model, id)
-            is Operations -> if (message.operations.isNotEmpty()) {
-                Operations(message.operations, id)
-            } else return true
+        val muxId = requireNotNull(message.muxId) {
+            "messages sent to Direct Store Muxer must have a muxId"
         }
-        return store.onProxyMessage(deMuxedMessage)
+        val (_, store) = getStore(muxId, message.id!!)
+        return store.onProxyMessage(message)
     }
 
-    /* internal */ suspend fun setupStore(referenceId: String): StoreRecord<Data, Op, T> {
+    /* internal */ suspend fun setupStore(muxId: String): StoreRecord<Data, Op, T> {
         val store = DirectStore.create<Data, Op, T>(
             StoreOptions(
-                storageKey = storageKey.childKeyWithComponent(referenceId),
+                storageKey = storageKey.childKeyWithComponent(muxId),
                 type = backingType
             )
         )
 
-        val id = store.on(callbackFactory(referenceId))
-
-        // Return a new Record and add it to our local stores, keyed by muxId.
-        return StoreRecord(id, store)
+        // Return a new Record with initially empty idSet
+        return StoreRecord(mutableSetOf<Int>(), store)
     }
 
-    @VisibleForTesting
-    suspend fun store(id: String) = storeMutex.withLock {
-        stores.getOrPut(id) {
-            setupStore(id)
+    /**
+     * Return the [StoreRecord] for a given muxId.
+     *
+     * If an id is provided, ensures that the id is registered to the [Store] of the [StoreRecord]
+     */
+    suspend fun getStore(muxId: String, id: Int? = null): StoreRecord<Data, Op, T> {
+        storeMutex.withLock {
+            val storeRecord = stores.getOrPut(muxId) {
+                setupStore(muxId)
+            }
+
+            // Register `id` with store if it is not already registered. It is necessary to register
+            // a callback per id because wehn teh Direct Store Muxer receives a message from the
+            // Direct Store, it utilizes the message id to determine which observer to redirect
+            // the message to.
+            if (id != null && !storeRecord.idSet.contains(id)) {
+                synchronized(proxyManager) {
+                    storeRecord.store.on(
+                        ProxyCallback { message ->
+                            proxyManager.getCallback(id)?.invoke(message.withMuxId(muxId))
+                        },
+                        id
+                    )
+                    storeRecord.idSet.add(id)
+                    callbackIdToMuxIdMap.getOrPut(id, { mutableSetOf() }).add(muxId)
+                }
+            }
+            return storeRecord
```

Comment:
I don't think this chunk needs to be within the storeMutex.withLock, right?

It's getting confusing having the `proxyManager` used as a lock, and having `storeMutex` as well, but not really a pattern in how they get used. You typically want to have a lock to maintain an invariant state.